ampersand-state.js

```diff
@@ -177,11 +178,12 @@ assign(Base.prototype, Events, {
             }
 
             isEqual = this._getCompareForType(def.type);
+            onChange = this._getOnChangeForType(def.type);
             dataType = this._dataTypes[def.type];
 
             // check type if we have one
             if (dataType && dataType.set) {
-                cast = dataType.set(newVal);
+                cast = dataType.set.call(this, newVal, options);
```

Comment:
Agreed, good point @latentflip. Making `this` available in `set` could enable problems like the ones fixed in this PR (side-effects where there shouldn't be). To that point, maybe the `compare` function should be unbound from `this` as well and encourage use of `onChange`?